### ISSUE TITLE
Iocage jails with dots fail to start

### ISSUE THREAD

**Issue by gronke** [iocage/libiocage#483]
Related to https://github.com/iocage/iocage/issues/588

When an iocage jail has dots in its name the jail command fails starting them.

**Comment by gronke** [comment]
Rather than replacing the characters libiocage should prevent jail names including dots.

**Comment by Corvan** [comment]
But why should dots not be allowed?
We use them in our blueprint-jail names and at least iocage works with them as well
https://github.com/iocage/iocage/pull/935

**Comment by gronke** [comment]
On FreeBSD jails are not allowed to have a dot (`.`) in their name identifier. A workaround would be to replace dots with a different character, but since all other characters are allowed, this would lead to the problem of conflicting names.

iocage/iocage#935 does replace `.` with `_`, but imagine somebody has a jail `foo.bar` and `foo_bar` on the same system.

I agree that having dots in jail names can be useful, but I don't see an elegant way to solve this without adding a layer of abstraction, such as for using hashes of the ioc jail name as identifier instead. This would come with the disadvantage of making it more complicated to use standard FreeBSD tooling to interface with jails.

We could decide to sacrifice another character that is no longer allowed in jail names to replace the dot. After some playing around a bit, it seems that a NULL byte does a decent job:

```console
# jail -c persist name="foo.bar" path="/rescue"
jail: jail "foo" not found
# jail -c persist name="foo\x00bar" path="/rescue"
# jail -r "foobar"
jail: "foobar" not found
# jail -r "foo\x00bar"
```

@Corvan do you think that is a good idea? It would only appear for internal identification of jails.

**Comment by urosgruber** [comment]
How about `@` character. At least one case I could use this is to differentiate between the versions or jails?

**Comment by igalic** [comment]
why, exactly, do we not allow `.` in jail names?

**Comment by gronke** [comment]
> How about @ character. At least one case I could use this is to differentiate between the versions or jails?

Yeah, we should be more permissive with allowed characters in jails. While dot characters would need to be replaced, other characters could have a negative on the CLI tool. The question is now if we should care or just allow every character, including different white space characters and ANSI control characters.

I just came up with an idea. The `/` (slash) character is used to differentiate between different root datasets used by ioc (defined in `/etc/rc.conf`), so I would like to make this our replacement for `.` in jail identifiers. Consumers of libioc (e.g. the CLI tool, daemons, etc) can then decide to whitelist certain characters, but libioc should not block any of those. This is similar to a discussion I had with @ kprovost on a BSD Meetup in in Essen about interface names - since that I'm using ANSI control characters to give my interfaces a beautiful color 😆 🤷‍♂
